### PR TITLE
Snapshot Datasource

### PR DESCRIPTION
**Pull request by doriac11**
# Description
- Create Datasource for Snapshot

# ISSUE TYPE
- Feature Pull Request

##### RESOURCE OR DATASOURCE NAME
Snapshot Datasource

##### OUTPUT Example 
```
+ {
              + expired                    = false
              + generation                 = null
              + linked                     = true
              + linked_storage_group       = [
                  + {
                      + background_define_in_progress = false
                      + defined                       = true
                      + linked_creation_timestamp     = "14:29:00 Thu, 29 Jun 2023 +0000"
                      + linked_volume_name            = "0017C"
                      + name                          = "ConnorsStorageGroup_LNK_SG_001"
                      + percentage_copied             = 0
                      + source_volume_name            = "0017B"
                      + track_size                    = 131072
                      + tracks                        = 28260
                    },
                ]
              + linked_storage_group_names = [
                  + "ConnorsStorageGroup_LNK_SG_001",
                ]
              + name                       = "HourlyDefault"
              + non_shared_tracks          = 0
              + num_source_volumes         = 1
              + num_storage_group_volumes  = 1
              + persistent                 = null
              + restored                   = false
              + secure_expiry_date         = null
              + snapid                     = 169496985601
              + source_volume              = [
                  + {
                      + capacity    = 1884
                      + capacity_gb = 3.44970703125
                      + name        = "0017B"
                    },
                ]
              + state                      = [
                  + "Established",
                ]
              + time_to_live_expiry_date   = "13:55:36 Fri, 30 Jun 2023 +0000"
              + timestamp                  = "13:55:36 Thu, 29 Jun 2023 +0000"
              + timestamp_utc              = null
              + tracks                     = 0
 },

```
##### ADDITIONAL INFORMATION
UT Test Output
```
$ TF_ACC=1 go test -count=1 -v -run TestAccSnapshotDataSource
 === RUN   TestAccSnapshotDataSource
    provider_test.go:90:
                        provider "powermax" {
                                username      = "smc"
                                password      = "smc"
                                endpoint      = "http://localhost:3000"
                                serial_number = "123"
                                pmax_version  = "100"
                                insecure      = true
                        }

config &{Host: Scheme: DefaultHeader:map[Accept:application/json; charset=utf-8 Authorization:Basic c21jOnNtYw== Content-Type:application/json; charset=utf-8 symid:123] UserAgent:terraform-powermax-provider/1.0.0 Debug:false Servers:[{URL:http://localhost:3000/univmax/restapi Description:http://localhost:3000/univmax/restapi Variables:map[]}] OperationServers:map[] HTTPClient:0xc0001faa20} header map[Accept:application/json; charset=utf-8 Authorization:Basic c21jOnNtYw== Content-Type:application/json; charset=utf-8 symid:123]config &{Host: Scheme: DefaultHeader:map[Accept:application/json; charset=utf-8 Authorization:Basic c21jOnNtYw== Content-Type:application/json; charset=utf-8 symid:123] UserAgent:terraform-powermax-provider/1.0.0 Debug:false Servers:[{URL:http://localhost:3000/univmax/restapi Description:http://localhost:3000/univmax/restapi Variables:map[]}] OperationServers:map[] HTTPClient:0xc0003bbc80} header map[Accept:application/json; charset=utf-8 Authorization:Basic c21jOnNtYw== Content-Type:application/json; charset=utf-8 symid:123]config &{Host: Scheme: DefaultHeader:map[Accept:application/json; charset=utf-8 Authorization:Basic c21jOnNtYw== Content-Type:application/json; charset=utf-8 symid:123] UserAgent:terraform-powermax-provider/1.0.0 Debug:false Servers:[{URL:http://localhost:3000/univmax/restapi Description:http://localhost:3000/univmax/restapi Variables:map[]}] OperationServers:map[] HTTPClient:0xc0001eaa20} header map[Accept:application/json; charset=utf-8 Authorization:Basic c21jOnNtYw== Content-Type:application/json; charset=utf-8 symid:123]config &{Host: Scheme: DefaultHeader:map[Accept:application/json; charset=utf-8 Authorization:Basic c21jOnNtYw== Content-Type:application/json; charset=utf-8 symid:123] UserAgent:terraform-powermax-provider/1.0.0 Debug:false Servers:[{URL:http://localhost:3000/univmax/restapi Description:http://localhost:3000/univmax/restapi Variables:map[]}] OperationServers:map[] HTTPClient:0xc0000e1aa0} header map[Accept:application/json; charset=utf-8 Authorization:Basic c21jOnNtYw== Content-Type:application/json; charset=utf-8 symid:123]config &{Host: Scheme: DefaultHeader:map[Accept:application/json; charset=utf-8 Authorization:Basic c21jOnNtYw== Content-Type:application/json; charset=utf-8 symid:123] UserAgent:terraform-powermax-provider/1.0.0 Debug:false Servers:[{URL:http://localhost:3000/univmax/restapi Description:http://localhost:3000/univmax/restapi Variables:map[]}] OperationServers:map[] HTTPClient:0xc00062f560} header map[Accept:application/json; charset=utf-8 Authorization:Basic c21jOnNtYw== Content-Type:application/json; charset=utf-8 symid:123]config &{Host: Scheme: DefaultHeader:map[Accept:application/json; charset=utf-8 Authorization:Basic c21jOnNtYw== Content-Type:application/json; charset=utf-8 symid:123] UserAgent:terraform-powermax-provider/1.0.0 Debug:false Servers:[{URL:http://localhost:3000/univmax/restapi Description:http://localhost:3000/univmax/restapi Variables:map[]}] OperationServers:map[] HTTPClient:0xc000496840} header map[Accept:application/json; charset=utf-8 Authorization:Basic c21jOnNtYw== Content-Type:application/json; charset=utf-8 symid:123]--- PASS: TestAccSnapshotDataSource (11.39s)
PASS
ok      terraform-provider-powermax/powermax/provider   12.588s
```

## Manual check for linter
```
$ make check
terraform fmt -recursive examples/
examples\data-sources\powermax_snapshot\data-source.tf
gofmt -s -w .
golangci-lint run --fix --timeout 5m
go vet
```
# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 80% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Unit tests
- [x] Acceptance tests

### CODE REVIEW

**Comment by terraform-providers-svc** [comment]
Can one of the admins verify this patch?

File: powermax/provider/powermax.env [hunk 1]
```diff
@@ -1,6 +1,6 @@
 TF_ACC=1
-POWERMAX_ENDPOINT=
-POWERMAX_USERNAME=
-POWERMAX_PASSWORD=
-POWERMAX_SERIAL_NUMBER=
+POWERMAX_ENDPOINT=http://localhost:3000
```
Comment (nikitajoshi1) [inline powermax/provider/powermax.env:2]: @doriac11 Remove user info.

File: powermax/models/snapshot.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+type filterTypeSnapshot struct {
+	Name types.String `tfsdk:"name"`
+}
+
+// SnapshotDetailModal describes the detail of snapshot data source.
+type SnapshotDetailModal struct {
+	// The name of the SnapVX snapshot.
+	Name types.String `tfsdk:"name"`
+	// The number of generation for the snapshot. Using snap IDs instead of generation numbers is preferred.
+	Generation types.Int64 `tfsdk:"generation"`
```
Comment (nikitajoshi1) [inline powermax/models/snapshot.go:38]: @doriac11 Can you remove extra spaces?

**Review by nikitajoshi1 — APPROVED** [review]

File: powermax/provider/powermax.env (hunk 1 above)
Comment (doriac11) [reply powermax/provider/powermax.env:2]: Ohh yes forgot to revert thanks Nikita!

File: powermax/models/snapshot.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	// The unique snap ID for the snapshot. This snap ID value does not change in the way generations can change when newer snapshots                         are created or terminated. Using snapset IDs instead of generation numbers is preferred.
```
Comment (nikitajoshi1) [inline powermax/models/snapshot.go:39]: @doriac11 Can you remove extra spaces?

File: examples/data-sources/powermax_snapshot/provider.tf [hunk 4]
```diff
@@ -0,0 +1,16 @@
+terraform {
+  required_providers {
+    powermax = {
+      source = "dell/powermax"
+    }
+  }
+}
+
+provider "powermax" {
```
Comment (shekhar-j) [inline examples/data-sources/powermax_snapshot/provider.tf:9]: @doriac11 aren't we keeping in separate file?

File: powermax/provider/snapshot_data_source.go [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+	}
+
+	listParam := d.client.PmaxOpenapiClient.ReplicationApi.GetStorageGroupSnapshots(ctx, d.client.SymmetrixID, plan.StorageGroup.Name.ValueString())
+	list, _, err := listParam.Execute()
+	if err != nil {
+		errStr := ""
+		message := helper.GetErrorString(err, errStr)
+		resp.Diagnostics.AddError(
+			"Error getting the list of snapshots",
+			constants.ReadSnapshots+"with error: "+message,
```
Comment (shekhar-j) [inline powermax/provider/snapshot_data_source.go:296]: @doriac11 you can pass this in helper call as errStr

**Review by shekhar-j — APPROVED** [review]

**Pull request merged**